fluent/migrate/transforms.py

```diff
@@ -92,42 +85,106 @@ def __call__(self, ctx):
+
+        Convert TextElements and StringExpressions into TextElements and join
+        adjacent ones.
+        '''
+
+        def get_text(element):
+            if isinstance(element, FTL.TextElement):
+                return element.value
+            elif isinstance(element, FTL.Placeable):
+                if isinstance(element.expression, FTL.StringExpression):
+                    return element.expression.value
+
+        joined = []
+        for current in elements:
+            current_text = get_text(current)
+            if current_text is None:
+                joined.append(current)
+                continue
+
+            previous = joined[-1] if len(joined) else None
+            if isinstance(previous, FTL.TextElement):
+                previous.value += current_text
+            elif len(current_text) > 0:
+                # Normalize to a TextElement
+                joined.append(FTL.TextElement(current_text))
+        return joined
+
+    @staticmethod
+    def preserve_whitespace(elements):
+        # Handle empty values
+        if len(elements) == 0:
+            return [FTL.Placeable(FTL.StringExpression(''))]
+
+        re_leading = re.compile(r'^(?P<whitespace>\s+)(?P<text>.*?)$')
+        re_trailing = re.compile(r'^(?P<text>.*?)(?P<whitespace>\s+)$')
+
+        def extract_whitespace(regex, element):
+            '''Extract leading or trailing whitespace from a TextElement.
+
+            Return a tuple of (Placeable, TextElement) in which the Placeable
+            encodes the extracted whitespace as a StringExpression and the
+            TextElement has the same amount of whitespace removed. The
+            Placeable with the extracted whitespace is always returned first.
+            '''
+            match = re.search(regex, element.value)
+            if match:
+                whitespace = match.group('whitespace')
+                empty_expr = FTL.Placeable(FTL.StringExpression(whitespace))
+                if whitespace == element.value:
+                    return empty_expr, None
+                else:
+                    return empty_expr, FTL.TextElement(match.group('text'))
             else:
-                pruned.extend(elems)
-        return pruned
+                return None, element
+
+        if isinstance(elements[0], FTL.TextElement):
+            ws, text = extract_whitespace(re_leading, elements[0])
+            elements[:1] = [ws, text]
+
+        if isinstance(elements[-1], FTL.TextElement):
+            ws, text = extract_whitespace(re_trailing, elements[-1])
+            elements[-1:] = [text, ws]
+
+        return [
+            element
+            for element in elements
+            if element is not None
+        ]
+
+    @staticmethod
+    def pattern_of(*elements):
+        elements = Transform.flatten_elements(elements)
+        elements = Transform.normalize_text_content(elements)
+        elements = Transform.preserve_whitespace(elements)
```

Comment:
Just sugar. The first `elements` is a generator, and then it switches to a list. I'm not opposed to the generator, but `preserve_whitespace` requires elements to actually be a list, so this feels a bit brittle?

Maybe just a comment?

Reply:
Good point. How about I rename `flatten_elements` to `chain_elements` which is closer to `itertools.chain`?

Reply:
I'll also move `flatten_elements`, `normalize_text_content` and `preserve_whitespace` to regular functions defined at the top of the file. I only really care about `Transform.pattern_of` being part of the official API because it may be useful in custom `foreach` functions passed to `PLURALS`.

Reply:
Or maybe it's better to keep it all namespaced under `Transform` after all? @Pike do you have an opinion?

Reply:
(Insert a joke about renaming to `smoosh_elements`.)

Reply:
Ah, I think I have a better idea. I'll push in a few.

Reply:
https://github.com/projectfluent/python-fluent/pull/54/commits/8f9bb30510fba7ec8017aa65a8366193ee69be5a is a refactor of `pattern_of`. I'm afraid rebasing this into the other three commits will be hard, so I might just land it as an independent commit.